Use an existing package of your choice to train and test a logistic regression model on the SPAMBASE dataset.

In [1]:
import pandas as pd
from ucimlrepo import fetch_ucirepo 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score

In [2]:
# import fetch dataset 
spambase = fetch_ucirepo(id=94) 
  
# data (as pandas dataframes) 
X = spambase.data.features 
y = spambase.data.targets 
  
# metadata 
#print(spambase.metadata) 
  
# variable information 
#print(spambase.variables) 

1. Split the original data into 75% for training and 25% for testing. Choose the training set at random. Train a logistic regression model on the training set and output the following **on the testing set**:
    1. **Confusion matrix**
    2. **Accuracy**, **Error**
    3. **Precision**, **Recall**, **F1 score**

In [7]:
# split data into training and test sets:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, shuffle=True) 

# scale the data: 
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test) 

# train logistic regression model:
model = LogisticRegression()
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)

/Users/pavithra/anaconda3/lib/python3.11/site-packages/sklearn/utils/validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [8]:
# outputs:
print("Confusion matrix: ", confusion_matrix(y_test, y_pred))
print("Accuracy score: ", accuracy_score(y_test, y_pred))
print("Precision: ", precision_score(y_test, y_pred))
print("Recall score: ", recall_score(y_test, y_pred))
print("F1 score: ", f1_score(y_test, y_pred))

Confusion matrix:  [[649  27]
 [ 62 413]]
Accuracy score:  0.9226759339704604
Precision:  0.9386363636363636
Recall score:  0.8694736842105263
F1 score:  0.9027322404371585


2. Print the coefficients of the features in the model. Which features contribute mostly to the prediction? Which ones are positively
correlated and which ones are negatively correlated with the SPAM class?

In [19]:
# print model intercept
print("Intercept: ", model.intercept_)

# print model coefficient in a table with corresponding feature
coeff = {
    'Feature': X.columns,
    'Coefficient': model.coef_.flatten()
    }

coeff_df = pd.DataFrame(coeff)
coeff_df

Intercept:  [-2.66842957]


,Feature,Coefficient
0,word_freq_make,-0.069564
1,word_freq_address,-0.170736
2,word_freq_all,0.066810
3,word_freq_3d,0.811284
4,word_freq_our,0.407512
5,word_freq_over,0.143471
6,word_freq_remove,0.851676
7,word_freq_internet,0.185421
8,word_freq_order,0.143223
9,word_freq_mail,0.040890


According to the table above, the features that contribute most to the prediction include word_freq_george, word_freq_hp, and word_freq_cs. The magnitude of the coefficients of these features are significantly greater than the magnitude of the coefficients of the other features, indicating that they contribute more to the prediction. 

In [20]:
# separate coefficients into positive and negative columns

positive = coeff_df[coeff_df['Coefficient'] > 0]['Feature'].reset_index(drop=True)
negative = coeff_df[coeff_df['Coefficient'] < 0]['Feature'].reset_index(drop=True)

pd.DataFrame({
    'Positive Coefficients': positive,
    'Negative Coefficients': negative
})


,Positive Coefficients,Negative Coefficients
0,word_freq_all,word_freq_make
1,word_freq_3d,word_freq_address
2,word_freq_our,word_freq_receive
3,word_freq_over,word_freq_will
4,word_freq_remove,word_freq_people
5,word_freq_internet,word_freq_hp
6,word_freq_order,word_freq_hpl
7,word_freq_mail,word_freq_george
8,word_freq_report,word_freq_lab
9,word_freq_addresses,word_freq_labs


Features that are positively correlated have a positive coefficient, such as word_freq_all, word_freq_3d and word_freq_3d as denoted in the first column of the table above. Features that are negatively correlated have negative coefficients, such as word_freq_make, word_freq_address, and word_freq_recieve in the second column from the table above.

3. Vary the decision threshold $T \in \{0.25,0.5,0.75,0.9\}$ and report for each value the model accuracy, precision, and recall. Comment on
how these metrics vary with the choice of threshold.

In [22]:
thresholds = [0.25, 0.5, 0.75, 0.9]
rows = []
for T in thresholds:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=T, random_state=42, shuffle=True) 
    # scale the data: 
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled  = scaler.transform(X_test)
    
    # train logistic regression model:
    model = LogisticRegression()
    model.fit(X_train_scaled, y_train)

    y_pred = model.predict(X_test_scaled)

    # report scores
    rows.append({
        "Decision Threshold": T,
        "Accuracy": accuracy_score(y_test, y_pred), 
        "Precision": precision_score(y_test, y_pred),
        "Recall": recall_score(y_test, y_pred),
    })

pd.DataFrame(rows)


/Users/pavithra/anaconda3/lib/python3.11/site-packages/sklearn/utils/validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/pavithra/anaconda3/lib/python3.11/site-packages/sklearn/utils/validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/pavithra/anaconda3/lib/python3.11/site-packages/sklearn/utils/validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/pavithra/anaconda3/lib/python3.11/site-packages/sklearn/utils/validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expec

,Decision Threshold,Accuracy,Precision,Recall
0,0.25,0.922676,0.938636,0.869474
1,0.50,0.918731,0.920046,0.870614
2,0.75,0.906114,0.900677,0.861771
3,0.90,0.899300,0.882500,0.860451


According to my table above, as the decision threshold increases from 0.25 to 0.90, the accuracy, precision, and recall scores all decrease respectively. The highest accuracy, precision and recall values are seen with the decision threshod of 0.25.